File: src/main/java/org/pyjjs/scheduler/core/data/HashSetDataSourceImpl.java
package org.pyjjs.scheduler.core.data;

import com.google.common.collect.Lists;
import org.pyjjs.scheduler.core.model.primary.IdentifiableObject;

import javax.annotation.Nonnull;
import java.util.*;

public class HashSetDataSourceImpl extends ObservableDataSource {

    private Set<IdentifiableObject> data;

    public HashSetDataSourceImpl() {
        data = new HashSet<>();
    }

    @Override
    public int size() {
        return data.size();
    }

    @Override
    public boolean isEmpty() {
        return data.isEmpty();
    }

    @Override
    public boolean contains(Object o) {
        return data.contains(o);
    }

    @Override
    @Nonnull
    public Iterator<IdentifiableObject> iterator() {
        return data.iterator();
    }

    @Override
    @Nonnull
    public Object[] toArray() {
        return data.toArray();
    }

    @Override
    @Nonnull
    public <T> T[] toArray(@Nonnull T[] a) {
        return data.toArray(a);
    }

    @Override
    public boolean add(IdentifiableObject entity) {
        boolean added = data.add(entity);
        if(added) {
            notifyAboutCreate(entity);
        } else {
            data.remove(entity);
            data.add(entity);
            notifyAboutUpdate(entity);
        }
        return added;

    }

    @Override
    public boolean remove(Object o) {
        boolean removed = data.remove(o);
        if(removed) {
            notifyAboutRemove((IdentifiableObject) o);
        }
        return removed;
    }

    @Override
    public boolean containsAll(@Nonnull Collection<?> c) {
        return data.containsAll(c);
    }

    @Override
    public boolean addAll(@Nonnull Collection<? extends IdentifiableObject> c) {
        boolean changed = false;
        for (IdentifiableObject identifiableObject : c) {
            changed |= add(identifiableObject);
        }
        return changed;
    }

    @Override
    public boolean removeAll(@Nonnull Collection<?> c) {
        boolean changed = false;
        for (Object o : c) {
            changed |= remove(o);
        }
        return changed;
    }

    @Override
    public boolean retainAll(@Nonnull Collection<?> c) {
        List<IdentifiableObject> removedData = Lists.newArrayList(data);
        removedData.removeAll(c);
        data.retainAll(c);
        for (IdentifiableObject entity : removedData) {
            notifyAboutRemove(entity);
        }
        return removedData.size() > 0;
    }

    @Override
    public void clear() {
        removeAll(data);
    }


}

File: src/main/kotlin/org/pyjjs/scheduler/core/data/HashSetDataSourceImpl.kt
package org.pyjjs.scheduler.core.data

import com.google.common.collect.Lists
import org.pyjjs.scheduler.core.model.IdentifiableObject
import java.util.*

class HashSetDataSourceImpl : ObservableDataSource() {

    private val data: MutableSet<IdentifiableObject>

    init {
        data = HashSet<IdentifiableObject>()
    }

    override val size: Int
        get() = data.size

    override fun contains(element: IdentifiableObject): Boolean {
        return data.contains(element)
    }

    override fun containsAll(elements: Collection<IdentifiableObject>): Boolean {
        return data.containsAll(elements)
    }

    override fun isEmpty(): Boolean {
        return data.isEmpty()
    }

    override fun iterator(): MutableIterator<IdentifiableObject> {
        return data.iterator()
    }

    override fun add(element: IdentifiableObject): Boolean {
        val added = data.add(element)
        if (added) {
            notifyAboutCreate(element)
        } else {
            data.remove(element)
            data.add(element)
            notifyAboutUpdate(element)
        }
        return added

    }

    override fun remove(element: IdentifiableObject): Boolean {
        val removed = data.remove(element)
        if (removed) {
            notifyAboutRemove(element)
        }
        return removed
    }

    override fun removeAll(elements: Collection<IdentifiableObject>): Boolean {
        var changed = false
        for (o in elements) {
            changed = changed or remove(o)
        }
        return changed
    }

    override fun retainAll(elements: Collection<IdentifiableObject>): Boolean {
        val removedData = Lists.newArrayList(data)
        removedData.removeAll(elements)
        data.retainAll(elements)
        for (entity in removedData) {
            notifyAboutRemove(entity)
        }
        return removedData.size > 0
    }

    override fun addAll(elements: Collection<IdentifiableObject>): Boolean {
        var changed = false
        for (identifiableObject in elements) {
            changed = changed or add(identifiableObject)
        }
        return changed
    }

    override fun clear() {
        removeAll(data)
    }

}
